### DISCUSSION TITLE
shouldUnregister doesn't unregister dynamic field in a nested component

### DISCUSSION THREAD

**Discussion by Muffinmann** [react-hook-form/react-hook-form#8095] [Q&A] [answered]
I'm listening to field changes through the 'onChange' event in a deeply nested component and dynamic fields will be triggered by certain values. However, they are still registered after unmount. I didn't dig very deep into the source code but I suppose this is because the "control._removeUnmounted" method in the useForm hook is not invoked (which is in the root component and React won't call that)?
By using 'useWatch' hook can walk around this problem... but I'm still wondering, is it a bug? or there's any mechanism I don't get about RHF?

Demo Sandbox: https://codesandbox.io/s/react-hook-form-basic-forked-obt7d5?file=/src/index.js
In this demo, when  one does the following actions:
1. choose 'y' for 'is a Developer' field. (then the email input is rendered)
2. choose 'n' (email input is unmounted)
3. click submit

Expected result:
email field should be unregistered.

Actual result:
email field still exists.

**Comment by Moshyfawn** [comment, marked as answer]
Can this be considered as a bug or a design limitation, @bluebill1049?

Tho, @Muffinmann, I don't see a reason to store the `isDeveloper` value in an external React state, since it's gonna be the same amount of renders as with watching the RHF value via `useWatch`

**Reply by bluebill1049** [reply]
https://react-hook-form.com/api/useform#shouldUnregister

> unmounted input will need to notify at either useForm, or useWatch's useEffect for hook form to verify input is unmounted from the DOM.

you can use `unregister` instead.

**Reply by Moshyfawn** [reply]
Ah, makes sense. Sry, being a bit inattentive

**Reply by bluebill1049** [reply]
no worries at all. thank you for the answer.

**Reply by aadlc** [reply]
Would you mind giving an example on  how you notify 
> at either `useForm`, or useWatch's `useEffect` for hook form to verify input is unmounted from the DOM.
